Add D0 case in the framework

zhangbiao-phy opened the pull request.

fgrosa commented on GetRawYieldsDplusDs.py:
```diff
@@ -117,12 +125,26 @@
 if not infile or not infile.IsOpen():
     print(f'ERROR: file "{args.inFileName}" cannot be opened! Exit!')
     sys.exit()
+infileref = TFile.Open(args.refFileName)
+if enableRef and not (infileref and infileref.IsOpen()):
+    print(f'ERROR: file "{args.refFileName}" cannot be opened! Exit!')
+    sys.exit()
```
```suggestion
if enableRef:
    infileref = TFile.Open(args.refFileName)
    if not (infileref and infileref.IsOpen()):
        print(f'ERROR: file "{args.refFileName}" cannot be opened! Exit!')
        sys.exit()
```

fgrosa commented on GetRawYieldsDplusDs.py:
```diff
@@ -117,12 +125,26 @@
+    print(f'ERROR: file "{args.refFileName}" cannot be opened! Exit!')
+    sys.exit()
 
+if enableRef:
```
```suggestion
```

fgrosa commented on GetRawYieldsDplusDs.py:
```diff
@@ -472,6 +502,14 @@
             else:
                 massFitter[iPt].IncludeSecondGausPeak(massDplus, False, fitConfig[cent]['SigmaSecPeak'][iPt], True)
+        if enableRef:
+            r_over_s = hMassForSig[iPt].Integral(hMassForSig[iPt].FindBin(massMin * 1.0001), hMassForSig[iPt].FindBin(massMax * 0.999))
+            r_over_s = hMassForRel[iPt].Integral(
+                                        hMassForRel[iPt].FindBin(massMin * 1.0001),
+                                        hMassForRel[iPt].FindBin(massMax * 0.999)) \
+                                        / r_over_s
+            massFitter[iPt].SetFixReflOverS(r_over_s)
```
```suggestion
            rOverS = hMassForSig[iPt].Integral(
                hMassForSig[iPt].FindBin(massMin * 1.0001),
                hMassForSig[iPt].FindBin(massMax * 0.999)
            )
            rOverS = hMassForRel[iPt].Integral(
                hMassForRel[iPt].FindBin(massMin * 1.0001),
                hMassForRel[iPt].FindBin(massMax * 0.999)
            ) / rOverS
            massFitter[iPt].SetFixReflOverS(rOverS)
```

fgrosa commented on filterdata/FilterTree4ML.py:
```diff
@@ -110,6 +110,8 @@
     elif 'Dstar' in channel:
         bitsForSel = {'bkg': [bitBkg], 'prompt_sig': [bitSignal, bitPrompt], 'FD_sig': [bitSignal, bitFD]}
+    elif 'D0' in channel:
+        bitsForSel = {'bkg': [bitBkg], 'prompt_sig': [bitSignal, bitPrompt], 'FD_sig': [bitSignal, bitFD], 'prompt_sig_refl': [bitSignal, bitPrompt, bitRefl], 'FD_sig_refl': [bitSignal, bitFD, bitRefl]}
```
```suggestion
        bitsForSel = {'bkg': [bitBkg], 'prompt_sig': [bitSignal, bitPrompt], 'FD_sig': [bitSignal, bitFD],
                      'prompt_sig_refl': [bitSignal, bitPrompt, bitRefl], 'FD_sig_refl': [bitSignal, bitFD, bitRefl]}
```

fgrosa reviewed (changes requested):
Hi @zhangbiao-phy! Many thanks for the implementation, I would just ask you to implement the minor comments I have and if possible also implement the `D0` case in the `c++` script for the raw-yield extraction [GetRawYieldsDplusDs.C](https://github.com/DmesonAnalysers/DmesonAnalysis/blob/master/GetRawYieldsDplusDs.C). Thanks a lot!

zhangbiao-phy commented:
> > Hi @zhangbiao-phy! Many thanks for the implementation, I would just ask you to implement the minor comments I have and if possible also implement the `D0` case in the `c++` script for the raw-yield extraction [GetRawYieldsDplusDs.C](https://github.com/DmesonAnalysers/DmesonAnalysis/blob/master/GetRawYieldsDplusDs.C). Thanks a lot!
> 
> Hi @fgrosa, I have changed the code based on your very nice suggestions. Let me know if you have more comments. Thanks a lot!

BTW, I have implemented D0 also in the C++ code

fcatalan92 commented on GetRawYieldsDplusDs.C:
```diff
@@ -175,19 +179,30 @@ int GetRawYieldsDplusDs(int cent, bool isMC, TString infilename, TString cfgfile
     //load inv-mass histos
     auto infile = TFile::Open(infilename.Data());
     if(!infile || !infile->IsOpen()) return -1;
+    auto infileref = TFile::Open(refFileName.Data());
+    if(!infileref || !infileref->IsOpen()) return -1;
```
This should be conditional on `enableRef`, otherwise the script would not work for the other particles

fcatalan92 commented on GetRawYieldsDplusDs.C:
```diff
@@ -401,7 +424,8 @@ int GetRawYieldsDplusDs(int cent, bool isMC, TString infilename, TString cfgfile
                 else {
                     massFunc = new TF1(Form("massFunc%d",iPt),DoublePeakSingleGaus,MassMin[iPt],MassMax[iPt],6);
-                    massFunc->SetParameters(hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massForFit,0.010,hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massDplus,0.010);
+                    if(particle==kDplus)massFunc->SetParameters(hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massForFit,0.010,hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massDplus,0.010);
+                    if(particle==kD0)massFunc->SetParameters(hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massForFit,0.010,hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massD0,0.010);
```
This part of the code is for the Ds specifically (which has a secondary peak due to the D+). It does not need to be modified for the D0

zhangbiao-phy replied:
yes, my mistake

fcatalan92 commented on GetRawYieldsDplusDs.C:
```diff
@@ -416,7 +440,9 @@ int GetRawYieldsDplusDs(int cent, bool isMC, TString infilename, TString cfgfile
                 else {
                     massFunc = new TF1(Form("massFunc%d",iPt),DoublePeakDoubleGaus,MassMin[iPt],MassMax[iPt],8);
-                    massFunc->SetParameters(hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massForFit,0.010,0.030,0.9,hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massDplus,0.010);
+                    if(particle==kDplus)massFunc->SetParameters(hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massForFit,0.010,0.030,0.9,hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massDplus,0.010);
+                    if(particle==kD0)massFunc->SetParameters(hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massForFit,0.010,0.030,0.9,hMassForFit[iPt]->Integral()*hMassForFit[iPt]->GetBinWidth(1),massD0,0.010);
+
```
See above

fcatalan92 commented on GetRawYieldsDplusDs.py:
```diff
@@ -17,6 +18,7 @@
 parser.add_argument('fitConfigFileName', metavar='text', default='config_Ds_Fit.yml')
 parser.add_argument('centClass', metavar='text', default='')
 parser.add_argument('inFileName', metavar='text', default='')
+parser.add_argument('--refFileName', metavar='text', default='')
```
I would put this option at the end of the parsed "needed" files, i.e., after outFileName

fcatalan92 commented on GetRawYieldsDplusDs.py:
```diff
@@ -1,6 +1,7 @@
 '''
-Script for fitting D+ and Ds+ invariant-mass spectra
+Script for fitting D+, D0 and Ds+ invariant-mass spectra
 run: python GetRawYieldsDsDplus.py fitConfigFileName.yml centClass inputFileName.root outFileName.root
+Add reflection file: python GetRawYieldsDsDplus.py fitConfigFileName.yml centClass inputFileName.root refFileName.root outFileName.root
```
Better to follow what done in, e.g., https://github.com/DmesonAnalysers/DmesonAnalysis/blob/master/ComputeDataDrivenCrossSection.py 
So you can just update the `run:` string

fcatalan92 commented on GetRawYieldsDplusDs.py:
```diff
@@ -339,8 +369,12 @@
                 massFunc = TF1(f'massFunc{iPt}', DoublePeakSingleGaus, massMin, massMax, 6)
-                massFunc.SetParameters(hMassForFit[iPt].Integral() * binWidth, massForFit, 0.010,
-                                       hMassForFit[iPt].Integral() * binWidth, massDplus, 0.010)
+                if(particleName == 'Dplus'):
+                    massFunc.SetParameters(hMassForFit[iPt].Integral() * binWidth, massForFit, 0.010,
+                                           hMassForFit[iPt].Integral() * binWidth, massDplus, 0.010)
+                if(particleName == 'D0'):
+                    massFunc.SetParameters(hMassForFit[iPt].Integral() * binWidth, massForFit, 0.010,
+                                           hMassForFit[iPt].Integral() * binWidth, massD0, 0.010)
```
Same comment as for the C++ macro

fcatalan92 commented on GetRawYieldsDplusDs.py:
```diff
@@ -353,8 +387,12 @@
                 massFunc = TF1(f'massFunc{iPt}', DoublePeakDoubleGaus, massMin, massMax, 8)
-                massFunc.SetParameters(hMassForFit[iPt].Integral() * binWidth, massForFit, 0.010, 0.030, 0.9,
-                                       hMassForFit[iPt].Integral() * binWidth, massDplus, 0.010)
+                if(particleName == 'Dplus'):
+                    massFunc.SetParameters(hMassForFit[iPt].Integral() * binWidth, massForFit, 0.010, 0.030, 0.9,
+                                           hMassForFit[iPt].Integral() * binWidth, massDplus, 0.010)
+                if(particleName == 'D0'):
+                    massFunc.SetParameters(hMassForFit[iPt].Integral() * binWidth, massForFit, 0.010, 0.030, 0.9,
+                                           hMassForFit[iPt].Integral() * binWidth, massD0, 0.010)
```
Same comment as for the C++ macro

fcatalan92 commented on GetRawYieldsDplusDs.py:
```diff
@@ -472,6 +510,17 @@
+                hMassForSig[iPt].FindBin(massMin * 1.0001),
+                hMassForSig[iPt].FindBin(massMax * 0.999)
+            )
+            rOverS = hMassForRel[iPt].Integral(
+                hMassForRel[iPt].FindBin(massMin * 1.0001),
+                hMassForRel[iPt].FindBin(massMax * 0.999)
+            ) / rOverS
```
You can use `/=`

fcatalan92 commented on GetRawYieldsDplusDs.py:
```diff
@@ -50,6 +52,10 @@
 ptMaxs = fitConfig[cent]['PtMax']
 fixSigma = fitConfig[cent]['FixSigma']
 fixMean = fitConfig[cent]['FixMean']
+if 'EnableRef' not in fitConfig[cent]:
+    enableRef = False
+else:
+    enableRef = fitConfig[cent]['EnableRef']
```
This conflict with the C++ macro, which crashes if no `EnableRef` entry is present in the yml file. 
For convention, we add all the options in the yml files. Please update all the configuration files for the fit for all the particles adding the new parameter.

zhangbiao-phy replied:
Ok

fcatalan92 commented on ProjectDplusDsSparse.py:
```diff
@@ -51,6 +51,7 @@
 infilenames = inputCfg['filename']
 if not isinstance(infilenames, list):
     infilenames = [infilenames]
+enableRef = inputCfg['enableRef']
```
Also here, please update all the yml files adding the new parameter

fcatalan92 commented on RunFullAnalysis.sh:
```diff
@@ -247,6 +247,8 @@ if $DoHFPtSpec; then
     Channel="kDsKKpi"
   elif [ ${Particle} == "Dplus" ]; then 
     Channel="kDplusKpipi"
+  elif [ ${Particle} == "D0" ]; then
+    Channel="kD0Kpi"
   fi
```
I think in this file you will need extra management for the raw yield extraction, where you have to pass the reflection file

fcatalan92 commented on configfiles/config_D0_pp_MC_sparse.yml:
```diff
@@ -0,0 +1,23 @@
+sparsenameAll: null
+sparsenamePrompt: fnSparseReco_fromC
+sparsenameFD: fnSparseReco_fromB
+enableRef: yes #no second peak for D+
```
Wrong comment

fcatalan92 commented on systematics/rawyields/RawYieldSystematics.cc:
```diff
@@ -77,13 +77,17 @@ int RawYieldSystematics(TString cfgFileName) {
     std::string mesonName = config["meson"].as<std::string>();    
     double massDs = TDatabasePDG::Instance()->GetParticle(431)->Mass();
     double massDplus = TDatabasePDG::Instance()->GetParticle(411)->Mass();
+    double massD0 = TDatabasePDG::Instance()->GetParticle(421)->Mass();
+
     double mass = -1;
     if(mesonName == "Ds")
         mass = massDs;
```
Also here you will have to manage the reflections

fcatalan92 reviewed (changes requested):
Hi @zhangbiao-phy, I have a few comments listed hereafter.
In general, I would advise you to test all the macros before committing them. It looks to me that you have modified scripts that you haven't used yet.
There is no need to do all the D0 integration together, it's fine if you commit the macros you used and modified bit by bit.

zhangbiao-phy commented:
> commit

Hi @fcatalan92 , Thanks for your comments. Indeed, I have run the code successfully on my laptop before committing. Maybe there are some things specific for the uncommitting code. for the systematic part, I will make another PR after I finish all the systematic studies. Let me know you have more comments. Thanks again!

fgrosa reviewed (approved):
Thanks @zhangbiao-phy!
